custom/icds/models.py

```diff
@@ -0,0 +1,163 @@
+
+from casexml.apps.case.models import CommCareCase
+from corehq.apps.userreports.expressions.getters import transform_date
+
+from dimagi.utils.decorators.memoized import memoized
+
+from .constants import *
+
+
+class ChildHealthCaseRow(object):
+    def __init__(self, case, snapshot_start_date, snapshot_end_date, forms_provider=[]):
+        """
+        Snapshot of the case in the duration specified by snapshot_start_date and snapshot_end_date
+
+        args:
+            case: case of type 'child_health'
+            snapshot_start_date: Start date of duration to snapshot the case on
+            snapshot_end_date: End date of duration to snapshot the case on.
+
+        dates can be date/datetime objects or a valid strings
+        """
+        if not isinstance(case, CommCareCase):
+            case = CommCareCase.wrap(case)
+        self.case = case
+        self.snapshot_start_date = transform_date(snapshot_start_date)
+        self.snapshot_end_date = transform_date(snapshot_end_date)
+        self.forms_provider = forms_provider
+
+    # needs caching across UCR build cycle
+    @property
+    @memoized
+    def forms(self):
+        if self.forms_provider:
+            return self.forms_provider
+        forms = []
+        for form_id in self.case.xform_ids:
+            forms.append(FormAccessors(domain).get_form(form_id))
+
+    def case_property(self, case, property, default=None):
+        prop = case.dynamic_case_properties().get(property, default)
+        if isinstance(prop, basestring) and prop.strip() == "":
```

Comment:
nit: this feels like a strange place for this logic. shouldn't `case_property` just return whatever property is specified in the case?